### ISSUE TITLE
Semver scheme

### ISSUE THREAD

**Issue by dobegor** [xanzy/go-gitlab#67]
I would really reconsider the decision about converting options to pointers. It's not a common way.
But please, follow the Semver [scheme](http://semver.org/). This was backward-incompatible change, but only minor part of the version number changed. My `glide update` just broke the CI pipeline :D 

I meant this commit: 4f1f8d2b5b095a5e718e3658602598890f4ec746.

**Comment by dobegor** [comment]
As to #29: we can still use non-pointer types for options, but provide pre-filled structs to `json.Marshal` to fill default values. That's a _much_ better option than converting all the options to pointers.

See this [example](http://stackoverflow.com/a/30445480).

**Comment by svanharmelen** [comment]
@dobegor I'm very sorry this update broke your CI pipeline! I know SemVer and we are actually using it as advertised. Please consider the following: http://semver.org/#spec-item-4 

So until we release version 1.0.0, any minor update can represent potentially breaking changes or otherwise big improvements/additions. While any patch updates represent either bug fixes or small new features. Both of which should be 100% backwards compatible.

Since you changed the title of this issue, may I assume you find the correct way to work with the various constants in the package?

And lastly about your suggestion for using pre-filled structs... I'm afraid that would not fix all use cases and would even introduce some new issues. For example say you would like to update the description or a project. In that case the API would accept these options:

```
Parameters:

id (required) - The ID of a project
name (optional) - project name
path (optional) - repository name for project
description (optional) - short project description
default_branch (optional)
issues_enabled (optional)
merge_requests_enabled (optional)
builds_enabled (optional)
wiki_enabled (optional)
snippets_enabled (optional)
container_registry_enabled (optional)
shared_runners_enabled (optional)
public (optional) - if true same as setting visibility_level = 20
visibility_level (optional)
public_builds (optional)
```

For this update call you would only like to set the required field and the description field, as you do not want to change anything else. In this case there is no way to know what all the other `defaults` should be, as in this case the `defaults` should be the currently configured values for that particular project.

I understand this design is a little more verbose and in a way doesn't feel right. But I believe this is the best way to make sure we can support all use cases to the best of our abilities. And I know for a fact there are more packages out there following this approach and that some blog posts are written on this topic as well.

Long story short, I'm not going to revert this change... Sorry!

Please let me know is your issue with using the constants is fixed as that is of course something that should work.

**Comment by dobegor** [comment]
Yeah, sorry, I was too quick to file the issue about constants - I've found a correct way using the `gitlab.VisibilityLevel` helper function.

Sorry, I didn't notice the major zero... more coffee needed indeed.
And yeah, filling the defaults of update calls may cause pain. Perhaps getting previous values first and filling the default struct with them would solve the problem, but that will do 2 API requests instead of one and may mislead user.

I am closing this now and sorry for bothering you before applying enough coffee :)

**Comment by svanharmelen** [comment]
@dobegor no worries :wink:
